### ISSUE TITLE
SQLite queries incorrectly formed using GUIDs.

### ISSUE THREAD

**Issue by DuaneTDAC** [linq2db/linq2db#1878]
I'm attempting to find a contact record based on a GUID.  
The GUID I'm searching for is: {e7a9e815-e4a5-4656-a45f-f15d27e3a265}

The code is:
```cs
Guid g = new Guid("e7a9e815-e4a5-4656-a45f-f15d27e3a265");
List<Contact> contacts = db.Contact.Where(x => x.CONTACTID == g)).ToList();
```
The SQLite query is coming out as below and returns 0 records.  
```sql
DECLARE @g0  -- Guid
SET     @g0 = Cast(x'15e8a9e7a5e45646a45ff15d27e3a265' as blob)
SELECT
	[x].[CONTACTID]
FROM
	[Contact] [x]
WHERE
	[x].[CONTACTID] = @g0

```
The same exact code works on an MS SQL database (exact copy of SQLite database) and produces the following SQL command:

```sql
DECLARE @g0 UniqueIdentifier -- Guid
SET     @g0 = 'e7a9e815-e4a5-4656-a45f-f15d27e3a265'
SELECT
	[x].[CONTACTID]
FROM
	[Contact] [x]
WHERE
	[x].[CONTACTID] = @g0

```
In reading, sqlite doesn't support DECLARE so I'm not sure about the DECLARE statement and why the GUID is being cast as BLOB and an odd value. 

Any help greatly appreciated.

**Comment by sdanyliv** [comment]
@DuaneTDAC, debug output contains only virtual parameters, just for understanding what are passed with query to the server. Many databases do not support variables in SQL, so treat it as debug information, real parameters are passed as a binary objects. 

Possible it is something new in SqLite which we do not handle. Try to rewrite query in the following way, which removes parameter.
```cs
List<Contact> contacts = db.Contact.Where(x => x.CONTACTID == Sql.ToSql(g))).ToList();
```
And then better to show SQL which woks correctly with your data.

**Comment by DuaneTDAC** [comment]
Okay, good to know.  Trying your suggestion changed the query to the below statement, but still returns 0 records.

SELECT
	[x].[CONTACTID]
FROM
	[ContactCustom] [x]
WHERE
	[x].[CONTACTID] = Cast(x'15e8a9e7a5e45646a45ff15d27e3a265' as blob)

If I change the query to below, and run with SQL query tool, I get the expected results.

SELECT
	[x].[CONTACTID]
FROM
	[ContactCustom] [x]
WHERE
	[x].[CONTACTID] = 'e7a9e815-e4a5-4656-a45f-f15d27e3a265'

The CONTACTID column is defined as 'uniqueidentifier' in the sqlite database.
I have also seen reference to setting BinaryGUID=True; in the connection string, but didn't make any difference.

**Comment by DuaneTDAC** [comment]
On a whim, I changed BinaryGUID=True to BinaryGUID=False in the connection string and now it's working correctly.  :)

**Comment by sdanyliv** [comment]
Anyway, it looks strange and we have to review that.

**Comment by DuaneTDAC** [comment]
3.28 is what I currently have installed.

**Comment by DuaneTDAC** [comment]
I have another project using EF6 and a sqlite database and it requires I have BinaryGuid=True.  Not sure why LinqToDB would be the opposite.

**Comment by sdanyliv** [comment]
@jogibear9988, can you handle this issue? Global configuration option would be prefer in 2.x version.

**Comment by sdanyliv** [comment]
@DuaneTDAC, just reviewed code and it looks like you can apply workaround just now, but it needs passing right converter to the mapping schema, like here:
https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/DataProvider/SQLite/SQLiteMappingSchema.cs#L21

```cs
var mappingSchema = new MappingSchema();
mappingSchema.SetValueToSqlConverter(typeof(Guid),     (sb,dt,v) => NewGuidToSql    (sb, (Guid)    v));

...
// just after DataConnection creation or in constructor
dc.AddMappingSchema(mappingSchema);
```

**Comment by DuaneTDAC** [comment]
Ok, thanks for the update!

**Comment by jogibear9988** [comment]
@sdanyliv should I do something?

That means we need a config option in sqlite provied if binary guids are used? But for me it seems BinaryGuid default is true, and we Dont use the option, so it should work without a change with binary guids.
I don't understand why it should work on his system using  BinaryGUID=False in connectionstring

**Comment by sdanyliv** [comment]
I do not understand what happened here either. So i suggest to create configuration option and handle both variants.
Ideally is parsing connection string and decide which strategy to use, but according to this issue looks like it is not a bullet proof way. So it’s why investigation is needed.

**Comment by jogibear9988** [comment]
i‘ll try to take a look

**Comment by jogibear9988** [comment]
I would not like ConnectionString parsing, what abou if odbc or a existing IDbConnection is used?

**Comment by sdanyliv** [comment]
Why not? SQLite is in-proc database. It is like getting setting from environment. Especially when connection string changes behavior.

It should be simple:
```cs
var values = connectionString.Split(';');
var keys = values.Select(v => v.Split('=')).ToDictionary(a => a[0], a => a.Length > 1 ? a[1] : null);
if (keys.TryGetValue("BinaryGUID", out var guidSetting)
{
   //
}
```

**Comment by sdanyliv** [comment]
Btw, if you can retrieve this information from SQLite via query - we can do that. Unfortunately I don't think that we have good mechanism for that.

**Comment by jogibear9988** [comment]
Yes, but on our system, linq2db does not create the dbconnection, we reuse one from the migrator

**Comment by sdanyliv** [comment]
So, just start from global setting, then we will return to this problem during Providers refactoring.

**Comment by jogibear9988** [comment]
global setting, what about different connections in one app?
first im looking if I could reproduce.

**Comment by DuaneTDAC** [comment]
I ran into another issue today.  As I previously mentioned, I set the connection string to BinaryGUID=False.  The data has been storing okay, lookups work etc.

Today I implemented BulkInsert for large imports.  The problem now is the GUIDs are being inserted as Binary, while single row inserts are Text.  So now lookups don't work correctly.

Any updates on this?

**Comment by Simson43** [comment]
> Why not? SQLite is in-proc database. It is like getting setting from environment. Especially when connection string changes behavior.
> 
> It should be simple:
> 
> ```cs
> var values = connectionString.Split(';');
> var keys = values.Select(v => v.Split('=')).ToDictionary(a => a[0], a => a.Length > 1 ? a[1] : null);
> if (keys.TryGetValue("BinaryGUID", out var guidSetting)
> {
>    //
> }
> ```

Is it working? how to set MappingSchema if I use EFCore?
Or there are other ways to resolve this issue?

**Comment by sdanyliv** [comment]
With EF Core currently you can correct `MappingSchema.Default` and only ONCE.
Will prepare better solution later.
